### PR TITLE
Add Chinese latex default support (refs: #3244 #3251)

### PR DESCRIPTION
**Pull request by cdarlint**
Subject: Add Chinese latex default support

### Purpose
- Add Chinese default latex support

### Detail
Problem Solved: Could not generate Chinese latex documents correctly by default
- Default xelatex+ctex in latex builder for Chinese when there is no user defined latex engine/latex_elements
- Detect ctex package for Chinese in latex writer and give warnings when not found. Did not force to ctex package and leave it configurable in conf.py for users
- Clear babel/polyglossia settings for Chinese in latex writer, as they are not suitable for Chinese language
- Default ctex package setting in conf.py generated by quickstart, otherwise default empty latex_elements in conf.py would override default settings in latex builder

### Relates
- #3244 initial commit
- #3251 @tk0miya refactored latex builder

### CODE REVIEW

File: sphinx/builders/latex.py [hunk 1]
```diff
@@ -268,13 +268,44 @@ def validate_config_values(app):
...
+                'howto': 'jreport'}
+    else:
+        return {}
+
+
+def default_latex_elements(config):
+    """ Better default latex_elements settings for specific languages. """
+    if config.language:
+        if config.language.split('_')[0] == 'zh':
+            return {'usepackages': '\\usepackage{ctex}'}
```
Comment (tk0miya) [inline sphinx/builders/latex.py:297]: As I said before, `usepackage` key is used only for internal. Don't use here.
It will be *OVERRIDED* if some extension uses `Sphinx.add_latex_package()` API.
We should use other key for this case.

File: sphinx/templates/quickstart/conf.py_t [hunk 2]
```diff
@@ -132,6 +132,11 @@ latex_elements = {
     # Latex figure (float) alignment
     #
     # 'figure_align': 'htbp',
+
+    # Chinese language default package ctex
```
Comment (tk0miya) [inline sphinx/templates/quickstart/conf.py_t:136]: Why is this needed? ctex is already loaded by default settings (see `default_latex_elements()`).

File: sphinx/writers/latex.py [hunk 3]
```diff
@@ -446,6 +446,18 @@ def __init__(self, document, builder):
             self.elements['multilingual'] = '%s\n\\setmainlanguage{%s}' % \
                 (self.elements['polyglossia'], self.babel.get_language())
 
+        # detect Chinese language, disable babel/polyglossia
```
Comment (tk0miya) [inline sphinx/writers/latex.py:449]: settings of babel and polyglossia are already finished here.
If you'd like to disable them, you should move these block to above.

File: sphinx/writers/latex.py [hunk 4]
```diff
@@ -446,6 +446,18 @@ def __init__(self, document, builder):
...
+        # give warnings when no common chinese package found in configuration
+        if builder.config.language:
+            if builder.config.language.split('_')[0] == 'zh':
+                self.elements['babel'] = ''
+                self.elements['polyglossia'] = ''
```
Comment (tk0miya) [inline sphinx/writers/latex.py:454]: This clears both keys always.
Did you mean all Chinese users could not use babel and polyglossia even if they configured intentionally?
I feel this is very strong restriction.

File: sphinx/writers/latex.py [hunk 5]
```diff
@@ -446,6 +446,18 @@ def __init__(self, document, builder):
...
+                if 'ctex' not in self.elements['usepackages'].lower():
```
Comment (tk0miya) [inline sphinx/writers/latex.py:455]: This is same above.
Isn't there any choices to build Chinese docs without ctex?

In this comment, https://github.com/sphinx-doc/sphinx/pull/3251#issuecomment-267512238, you said CJKutf8 package allows to build Chinese docs.
This warns to the such users. Is this intended?

**Review by tk0miya — CHANGES_REQUESTED** [review]
You made this PR to stable branch.
But I feel this makes incompatibilities as you said to Chinese users who already customized their project.

>different combinations of latex compiler/chinese support package/document settings may not work together.
>only several designed combinations of configuration works for chinese latex documents

So I think this should move to master branch, and we should announce about this new default settings.

File: sphinx/writers/latex.py (hunk 4 above)
Comment (cdarlint) [reply sphinx/writers/latex.py:454]: yes I mean it. There is no possible to use babel/polyglossia with chinese.

File: sphinx/writers/latex.py (hunk 5 above)
Comment (cdarlint) [reply sphinx/writers/latex.py:455]: Yes it is intended. ctex is the usually choice in most cases.
Chinese user would always get an warning by babel because babel does not support Chinese.
So I think it is usual that common user would use ctex and advanced user would understand what this warning means.
If your opinion is no warning in proper working cases, I'll think about it more deep

File: sphinx/templates/quickstart/conf.py_t (hunk 2 above)
Comment (cdarlint) [reply sphinx/templates/quickstart/conf.py_t:136]: because the empty dictionary by default overrides `default_latex_elements()`.
Another option would be comment whole `latex_elements={ }` in conf.py_t

File: sphinx/builders/latex.py (hunk 1 above)
Comment (cdarlint) [reply sphinx/builders/latex.py:297]: Maybe we should add a dedicated key for this case then

File: sphinx/writers/latex.py (hunk 3 above)
Comment (cdarlint) [reply sphinx/writers/latex.py:449]: OK I'll check the code later.

And about the PR to stable branch. In this version the latex builder/writer does not overrides conf.py so I think it is compatible.
I'll think about where to put `usepackage` and if I add a new conf key then maybe it is better to PR to master branch

File: sphinx/writers/latex.py (hunk 5 above)
Comment (cdarlint) [reply sphinx/writers/latex.py:455]: maybe we can clear this warning and add some notes to user manual?
and, is it good to clear babel warnings for chinese?

**Comment by cdarlint** [comment]
MEMO
issues for now (change requested)
1 do not use key "usepackage"
2 default in conf.py_t or default in latex builder/writer?
3 give warning if not using ctex? or place note in documentation? and, other possible compile conf available for zh_CN (not only chs, but also cht for TW/HK/SG/etc.)

also see #1119 
and be aware of new texlive-2013 as in #3070

**Comment by tk0miya** [comment]
Sorry, I just removed `stable` branch last night. As a result, github closes this automatically. It was not intended. so I reopened this again and changed target branch to `1.7` (we decided to rename stable branch to versioned one).
Thanks,
